### PR TITLE
DR2-1964 Update byIdentifiers and refactor tests

### PR DESCRIPTION
**Pull request by MancunianSam**
The entitiesByIdentifier method in the Preservica client has been
updated with a slightly different name and signature. Fixing the tests
to cope with this was really difficult because of all the mocking we've
got everywhere so I've replaced this in the three lambdas that call this
method.

The tests won't pass until https://github.com/nationalarchives/da-aws-clients/pull/318 is merged and deployed but this is ready for review otherwise.

### CODE REVIEW

**Review by techncl — COMMENTED** [review]
Looks good! Just a few comments.

We seem to have the same methods in multiple ExternalServicesTestUtils files; could we make it one file in the `utils` folder or something?

File: scala/lambdas/ingest-asset-reconciler/src/main/scala/uk/gov/nationalarchives/ingestassetreconciler/Lambda.scala [hunk 1]
```diff
@@ -190,7 +188,12 @@ object Lambda {
 
   case class AssetMessage(messageId: UUID, parentMessageId: Option[UUID] = None, executionId: Option[String])
 
-  case class StateOutput(wasReconciled: Boolean, reason: String, assetId: UUID)
+  enum FailureReason:
+    case NoContentObjects, TitleChecksumMismatch
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-asset-reconciler/src/main/scala/uk/gov/nationalarchives/ingestassetreconciler/Lambda.scala:192]: I'm going to leave it. There's enough context generally to know that we're probably not generating a checksum for the title. I also want to stop having names that are over 3 words long generally and ones with conjunctions more specifically as it's making it quite hard to scan the code with names that are too long.

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala [hunk 2]
```diff
@@ -107,24 +108,19 @@ class Lambda extends LambdaRunner[StepFnInput, Unit, Config, Dependencies] {
...
-        )
-        .map {
-          _.collect { case (value, Some(entity)) =>
-            (value, entity)
-          }.toMap
-        }
+      for {
+        entityMap <- dependencies.entityClient.entitiesPerIdentifier(identifiers)
+        multipleEntries = entityMap.filter { case (value, entities) => entities.length > 1 }
+        _ <- IO.raiseWhen(multipleEntries.nonEmpty)(new Exception(s"There is more than 1 entity with the same SourceIDs ${multipleEntries.keys.map(_.value).mkString(" ")}"))
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:119]: I've changed the error message. I'll leave them just seperated by spaces though, commas might just add noise to the message. We'll try it without them anyway.

File: scala/lambdas/ingest-find-existing-asset/src/test/scala/uk/gov/nationalarchives/ingestfindexistingasset/testUtils/ExternalServicesTestUtils.scala [hunk 3]
```diff
@@ -1,183 +1,159 @@
...
+    override def writeItems[T](tableName: String, items: List[T])(using format: DynamoFormat[T]): IO[List[BatchWriteItemResponse]] = notImplemented
+
+    override def queryItems[U](tableName: String, requestCondition: RequestCondition, potentialGsiName: Option[String])(using returnTypeFormat: DynamoFormat[U]): IO[List[U]] =
+      notImplemented
+
+    override def getItems[T, K](primaryKeys: List[K], tableName: String)(using returnFormat: DynamoFormat[T], keyFormat: DynamoFormat[K]): IO[List[T]] =
+      IO.raiseWhen(dynamoError)(ResourceNotFoundException.builder.message(s"$tableName not found").build) >> ref.get.map(_.map(_.asInstanceOf[T]))
+
+    override def updateAttributeValues(dynamoDbRequest: DADynamoDBClient.DADynamoDbRequest): IO[Int] = ref
+      .update { existing =>
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-find-existing-asset/src/test/scala/uk/gov/nationalarchives/ingestfindexistingasset/testUtils/ExternalServicesTestUtils.scala:90]: I think what I was going for here is that for every test client, for every lambda the state is passed in with a `Ref` called `ref` and the existing state is called `existing` so it's all uniform.
It means that all the test clients look roughly the same and you don't have to think about names when you make a new one.
I'm not 100% sure this is the right approach but I'm going to stick with it for now.

File: scala/lambdas/ingest-find-existing-asset/src/test/scala/uk/gov/nationalarchives/ingestfindexistingasset/testUtils/ExternalServicesTestUtils.scala [hunk 4]
```diff
@@ -1,183 +1,159 @@
...
+
+      override def streamBitstreamContent[T](stream: capabilities.Streams[Fs2Streams[IO]])(url: String, streamFn: stream.BinaryStream => IO[T]): IO[T] = notImplemented
+
+      override def entitiesUpdatedSince(dateTime: ZonedDateTime, startEntry: Int, maxEntries: Int): IO[Seq[Entity]] = notImplemented
+
+      override def entityEventActions(entity: Entity, startEntry: Int, maxEntries: Int): IO[Seq[DataProcessor.EventAction]] = notImplemented
+
+      override def entitiesPerIdentifier(identifiers: Seq[PreservicaIdentifier]): IO[Map[PreservicaIdentifier, Seq[Entity]]] =
+        IO.raiseWhen(apiError)(new Exception("API has encountered an error")) >>
+          ref.get.map { existing =>
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-find-existing-asset/src/test/scala/uk/gov/nationalarchives/ingestfindexistingasset/testUtils/ExternalServicesTestUtils.scala:130]: As above

File: scala/lambdas/ingest-find-existing-asset/src/test/scala/uk/gov/nationalarchives/ingestfindexistingasset/testUtils/ExternalServicesTestUtils.scala [hunk 5]
```diff
@@ -1,183 +1,159 @@
...
+            existing
+              .flatMap(_.identifiers)
+              .map { value =>
+                value -> existing.filter(_.identifiers.contains(value)).map(_.entity)
+              }
+              .toMap
+              .view
+              .filterKeys(identifiers.contains)
+              .toMap
+          }
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-find-existing-asset/src/test/scala/uk/gov/nationalarchives/ingestfindexistingasset/testUtils/ExternalServicesTestUtils.scala:140]: The first one is neater, thanks.

**Comment by MancunianSam** [comment]
I most definitely don't want a common method in the `utils` package. Reason is that most of then have slightly different requirements, different input objects etc. and it we try to make a generic one, it's going to end up very complicated and this is all about simplifying things. So I can live with the duplication as a trade for simplicity.

What I will probably end up doing is creating a trait in the utils packge for every trait we override with `notImplemented` for all the methods and then we can override that one which should shorten things considerably but I want to get all the changes in first before I do that.

**Review by techncl — APPROVED** [review]
👍

**Pull request merged**